freestyle/shared/src/main/scala/freestyle/free.scala

```diff
@@ -11,208 +11,163 @@ class free extends StaticAnnotation {
 
 object free {
 
-  def impl(c: whitebox.Context)(annottees: c.Expr[Any]*): c.universe.Tree = {
+  def impl(c: blackbox.Context)(annottees: c.Expr[Any]*): c.universe.Tree = {
     import c.universe._
     import internal.reificationSupport._
 
     def fail(msg: String) = c.abort(c.enclosingPosition, msg)
 
     def gen(): Tree = annottees match {
-      case List(Expr(cls: ClassDef)) => genModule(cls)
+      case List(Expr(cls: ClassDef)) =>
+        val effectTrait @ ClassDef(clsMods, _, _, _) = cls.duplicate
+        if (clsMods.hasFlag(Flag.TRAIT | Flag.ABSTRACT)) {
+          val effectObject = mkEffectObject(effectTrait)
+          q"""
+            $effectTrait
+            $effectObject
+          """
+        } else
+          fail(s"@free requires trait or abstract class")
       case _ =>
         fail(
           s"Invalid @free usage, only traits and abstract classes without companions are supported")
     }
 
-    def genModule(cls: ClassDef) = {
-      val userTrait @ ClassDef(clsMods, clsName, clsParams, clsTemplate) = cls.duplicate
-      if (!clsMods.hasFlag(Flag.TRAIT | Flag.ABSTRACT))
-        fail(s"@free requires trait or abstract class")
-      mkCompanion(clsName.toTermName, clsTemplate.filter {
-        case _: DefDef => true
-        case _         => false
-      }, clsParams, userTrait)
-    }
+    // OP is the name of the Root trait of the Effect ADT
+    lazy val OP = TypeName("T")
+    // MM Is the target of the Interpreter's natural Transformation
+    lazy val MM = TypeName("MM")
+    // LL is the target of the Lifter's Injection
+    lazy val LL = TypeName("LL")
+    // LIFT is the name of the Lifter's Injector Class: no sense in separate name
+    lazy val LIFT = TypeName("Lift")
```

Comment:
I think referring to the smart constructors as `Lifter` or `Lift` is confusing since we already have https://github.com/47deg/freestyle/blob/master/freestyle/shared/src/main/scala/freestyle/Lift.scala.